# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1736105017 -3600
# Parent  2ff37f27bb11a501bbc9b16dc66b29906ebd3899
Bug 1767190/1801464 - Fix issue in nsParseNewMailState::MoveIncorporatedMessage().

Fix long-standing issue in nsParseNewMailState::MoveIncorporatedMessage():
Header is announced before message is present in the store.
Fixes: Virtual folder based on body search not working when message received via POP moved by filter.

diff --git a/mailnews/local/src/nsParseMailbox.cpp b/mailnews/local/src/nsParseMailbox.cpp
--- a/mailnews/local/src/nsParseMailbox.cpp
+++ b/mailnews/local/src/nsParseMailbox.cpp
@@ -2173,10 +2173,13 @@ nsresult nsParseNewMailState::MoveIncorp
   NS_WARNING_ASSERTION(destMailDB && NS_SUCCEEDED(rv),
                        "failed to open mail db parsing folder");
   nsCOMPtr<nsIMsgDBHdr> newHdr;
 
+  // We don't add the header to the DB just yet, since the data isn't in the
+  // store and the storeToken is stale. Any listener won't be able to retrieve
+  // the message. Try a virtual folder based on body search.
   if (destMailDB)
-    rv = destMailDB->CopyHdrFromExistingHdr(m_new_key, mailHdr, true,
+    rv = destMailDB->CopyHdrFromExistingHdr(m_new_key, mailHdr, false,
                                             getter_AddRefs(newHdr));
   if (NS_SUCCEEDED(rv) && !newHdr) rv = NS_ERROR_UNEXPECTED;
 
   if (NS_FAILED(rv)) {
@@ -2185,8 +2188,10 @@ nsresult nsParseNewMailState::MoveIncorp
     rv = AppendMsgFromStream(inputStream, newHdr, destIFolder);
     if (NS_FAILED(rv))
       destIFolder->ThrowAlertMsg("filterFolderWriteFailed", msgWindow);
   }
+  if (destMailDB && NS_SUCCEEDED(rv))
+    rv = destMailDB->AddNewHdrToDB(newHdr, true);
 
   if (NS_FAILED(rv)) {
     if (destMailDB) destMailDB->Close(true);
 
